torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1117,6 +1119,207 @@ def arg_extractor(combine_fn, input, dim):
+                        if ini.size is not None
+                        else tuple(
+                            BuiltinVariable(getattr)
+                            .call_function(
+                                tx, [ini, ConstantVariable.create("shape")], {}
+                            )
+                            .items
+                        ),
+                    ),
+                ),
+                kwargs={
+                    "dtype": SourcelessBuilder.create(tx, ini.dtype),
+                    "device": SourcelessBuilder.create(tx, ini.device),
+                    "requires_grad": SourcelessBuilder.create(tx, ini.requires_grad),
+                },
+            )
+            for ini in init.items
+        ]
+        sub_args_inp_shapes = [
+            tuple(
+                sh if i != dim_fake else 1
+                for i, sh in enumerate(
+                    # get_fake_value(inp.as_proxy().node, tx).size()
+                    tuple(
+                        BuiltinVariable(getattr)
+                        .call_function(tx, [inp, ConstantVariable.create("shape")], {})
+                        .items
+                    )
+                )
+            )
+            for inp in input.items
+        ]
+        sub_args_inp = [
+            inp.call_method(
+                tx,
+                "new_empty",
+                args=(SourcelessBuilder.create(tx, inp_sh),),
+                kwargs={
+                    "dtype": SourcelessBuilder.create(tx, inp.dtype),
+                    "device": SourcelessBuilder.create(tx, inp.device),
+                    "requires_grad": SourcelessBuilder.create(tx, inp.requires_grad),
+                },
+            )
+            for inp, inp_sh in zip(input.items, sub_args_inp_shapes)
+        ]
+        sub_args = sub_args_init + sub_args_inp
+        (
+            (combine_result, combine_treespec),
+            combine_graph,
+            combine_lifted_freevars,
+        ) = speculate_subgraph(
+            tx,
+            combine_fn,
+            sub_args,
+            sub_kwargs={},
+            description="scan_combine_fn",
+            source_target=self.value,
+            set_subgraph_inputs="flatten_manual",
+        )
+
+        if combine_lifted_freevars:
+            unimplemented(
+                f"Combine fn had unexpected freevars: {combine_lifted_freevars}"
+            )
+
+        if any(cr.python_type() != list for cr in combine_result.items):
+            unimplemented(
+                f"Expected combine_fn to return a list if tensor but got {combine_result.python_type()}",
+            )
+
+        input_proxy = input.as_proxy()
+        init_proxy = init.as_proxy()
+        combine_carry_proxy = combine_result.items[0].as_proxy()
+        combine_output_proxy = combine_result.items[1].as_proxy()
+        for carry, out, inp_proxy, ini_proxy, s_args in zip(
+            combine_carry_proxy, combine_output_proxy, input_proxy, init_proxy, sub_args
```

Comment:
combine_carry_proxy and combine_output_proxy doesn't have to have same length right? because they're different pytree structures. I think we might want to split checks according to same pytree length.

Reply:
I split the checks into checks on init and carry and then checks on input